src/charm.py

```diff
@@ -206,32 +196,30 @@ def _on_certificates_relation_broken(self, event: EventBase) -> None:
         self._delete_certificate()
         self.unit.status = BlockedStatus("Waiting for certificate relation")
 
-    def _on_certificates_relation_joined(self, event: EventBase) -> None:
-        """Generates CSR and requests new certificate."""
-        if not self._container.can_connect():
-            event.defer()
-            return
-        if not self._private_key_is_stored():
-            event.defer()
-            return
-        if self._certificate_is_stored():
-            return
+    def _get_current_provider_certificate(self) -> str | None:
+        """Compares the current certificate request to what is in the interface.
 
-        self._request_new_certificate()
+        Returns the current valid provider certificate if present
+        """
+        csr = self._get_stored_csr()
+        for provider_certificate in self._certificates.get_assigned_certificates():
+            if provider_certificate.csr == csr:
+                return provider_certificate.certificate
+        return None
 
-    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
-        """Pushes certificate to workload and configures workload."""
-        if not self._container.can_connect():
-            event.defer()
-            return
-        if not self._csr_is_stored():
-            logger.warning("Certificate is available but no CSR is stored")
-            return
-        if event.certificate_signing_request != self._get_stored_csr():
-            logger.debug("Stored CSR doesn't match one in certificate available event")
-            return
-        self._store_certificate(event.certificate)
-        self._configure_sdcore_smf(event)
+    def _update_certificate(self, provider_certificate) -> bool:
```

Comment:
This method does more than it's description. It detects if certificate update is required, updates and returns if certificate is changed. We prefer to give single responsibility for each method.

Reply:
It updates the certificate if no certificate is currently present or it is doesn't match, returning `True` if it was updated of `False` otherwise; we do the same in `_update_config_file` method. I don't think splitting them into different methods improves the code readability, as even more logic would need to be moved to the main `_configure` callback. What do you think?

Reply:
IMO, the logic in the `_configure` callback could be improved.

Reply:
We can improve it by using `collect_status`. That would be easier.